sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOTest.java

```diff
@@ -492,6 +498,45 @@ public void testReadAvroSpecificRecordsWithConfluentSchemaRegistry() {
     p.run();
   }
 
+  @Test
+  public void testReadValuesAsAvroGenericRecordsWithSqlQuery() {
+    int numElements = 100;
+    String topic = "my_topic";
+    String schemaRegistryUrl = "mock://my-scope-name";
+    String valueSchemaSubject = topic + "-value";
+
+    Schema schema = AvroUtils.getSchema(GenericRecord.class, AVRO_SCHEMA);
+    List<Row> outputs = new ArrayList<>();
+    for (int i = 0; i < numElements; i++) {
+      outputs.add(Row.withSchema(schema).addValues("ValueName" + i, i, "color" + i).build());
+    }
+
+    PTransform<PBegin, PCollection<GenericRecord>> reader =
+        KafkaIO.<Integer, GenericRecord>read()
+            .withBootstrapServers("localhost:9092")
+            .withTopic(topic)
+            .withKeyDeserializer(IntegerDeserializer.class)
+            .withValueDeserializer(mockDeserializerProvider(schemaRegistryUrl, valueSchemaSubject))
+            .withConsumerFactoryFn(
+                new ConsumerFactoryFn(
+                    ImmutableList.of(topic),
+                    1,
+                    numElements,
+                    OffsetResetStrategy.EARLIEST,
+                    i -> ByteBuffer.wrap(new byte[4]).putInt(i).array(),
+                    new ValueAvroSerializableFunction(topic, schemaRegistryUrl)))
+            .withMaxNumRecords(numElements)
+            .withAvroSchemaValues();
```

Comment:
You can replace this `.withoutMetadata()` chained with `Keys.create()` and/or `Values.create()` to extract the respective to be queried after.